apps/backend/src/libs/modules/token/base-token.module.ts

```diff
@@ -0,0 +1,33 @@
+import { JWTPayload, jwtVerify, JWTVerifyResult, SignJWT } from "jose";
+
+type Constructor = {
+	algorithm: string;
+	expirationTime: string;
+	secret: Uint8Array;
+};
+
+class BaseToken<T extends JWTPayload> {
```

Comment:
I think we should not extend the payload token type from "jose" in case we need to change the JWT library.

So the decode method should return `Promise<T>`
```ts
import { jwtVerify, JWTVerifyResult, SignJWT } from "jose";

class BaseToken<PayloadT> {
  /* ... */
  public async createToken(payload: PayloadT): Promise<string> {
    return await new SignJWT(payload)
      .setProtectedHeader({ alg: this.algorithm })
      .setIssuedAt()
      .setExpirationTime(this.expirationTime)
      .sign(this.secret);
  }

  public async decode(token: string): Promise<PayloadT> {
    const { payload } = await jwtVerify<PayloadT>(token, this.secret);

    return payload;
  }
}
```

Reply:
I don't agree. Currently decode returns `Promise<JWTVerifyResult<T>>`, which is a generic, where T is our custom type